Sources/Core/AWSClientRuntime/EventStream/AWSMessageDecoder.swift

```diff
@@ -0,0 +1,96 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import AwsCommonRuntimeKit
+import ClientRuntime
+
+extension AWSEventStream {
+
+    /// AWS implementation of `MessageDecoder` for decoding event stream messages
+    /// Note: This is class because struct does not allow using `self` in closure
+    ///      and we need to use `self` to access `messageBuffer` per CRT API.
```

Comment:
This is fine, but if we did want this to be a struct, we can create a box to store the mutable properties.
Additionally, this would allow us to convert `decoder` to a `let`, since we won't need to use `self` in the closures.

```swift
struct AWSMessageDecoder: MessageDecoder {
    private class Store {
        var messageBuffer: [EventStream.Mesage] = []
        // etc...
    }
    
    private let store: Store
    
    public init() {
        let store = Store()
        let logger = SwiftLogger(label: "AWSMessageDecoder")
        
        self.decoder = EventStreamMessageDecoder(
            onPayloadSegment: { payload, finalSegment in 
                logger.debug("...")
                store.decodedPayload.append(payload)
            },
            // etc...
         )
         
         self.store = store
         self.logger = logger
    }
}
```

Reply:
for now leaving it as it is, because we don't anything wrapping the implementation inside the class.

Reply:
The one nice thing we get is that `decoder` no longer needs to be a variable and no longer needs to be an optional